ui/admin/app/components/model-providers/ModelProviderForm.tsx

```diff
@@ -0,0 +1,167 @@
+import { zodResolver } from "@hookform/resolvers/zod";
+import { useFieldArray, useForm } from "react-hook-form";
+import { z } from "zod";
+
+import { ModelProviderConfig } from "~/lib/model/modelProviders";
+
+import { TypographyH4 } from "~/components/Typography";
+import {
+    NameDescriptionForm,
+    ParamFormValues,
+} from "~/components/composed/NameDescriptionForm";
+import { ControlledInput } from "~/components/form/controlledInputs";
+import { useModelProviders } from "~/components/model-providers/ModelProviderContext";
+import { Button } from "~/components/ui/button";
+import { Form } from "~/components/ui/form";
+import { Separator } from "~/components/ui/separator";
+import { useAsync } from "~/hooks/useAsync";
+
+const formSchema = z.object({
+    requiredConfigParams: z.array(
+        z.object({
+            name: z.string().min(1, {
+                message: "Name is required.",
+            }),
+            value: z.string().min(1, {
+                message: "This field is required.",
+            }),
+        })
+    ),
+    additionalConfirmParams: z.array(
+        z.object({
+            name: z.string(),
+            description: z.string(),
+        })
+    ),
+});
+
+export type ModelProviderFormValues = z.infer<typeof formSchema>;
+
+const getInitialRequiredParams = (
+    requiredParameters: string[],
+    parameters: ModelProviderConfig
+): ModelProviderFormValues["requiredConfigParams"] =>
+    requiredParameters.map((requiredParameterKey) => ({
+        name: requiredParameterKey,
+        value: parameters[requiredParameterKey] ?? "",
+    }));
+
+const getInitialAdditionalParams = (
+    requiredParameters: string[],
+    parameters: ModelProviderConfig
+): ParamFormValues["params"] => {
+    const defaultEmptyParams = [{ name: "", description: "" }];
+
+    const requiredParameterSet = new Set(requiredParameters);
+    const additionalParams = Object.entries(parameters).filter(
+        ([key]) => !requiredParameterSet.has(key)
+    );
+    return additionalParams.length === 0
+        ? defaultEmptyParams
+        : additionalParams.map(([key, value]) => ({
+              name: key,
+              description: value,
+          }));
+};
+
+export function ModelProviderForm({
+    modelProviderId,
+    onSuccess,
+    parameters,
+    requiredParameters,
+}: {
+    modelProviderId: string;
+    onSuccess: (config: ModelProviderConfig) => void;
+    parameters: ModelProviderConfig;
+    requiredParameters: string[];
+}) {
+    const { configureModelProvider } = useModelProviders();
+
+    const form = useForm<ModelProviderFormValues>({
+        resolver: zodResolver(formSchema),
+        mode: "onChange",
```

Comment:
nit: this form is pretty simple at it's core (just checks for required fields). I'd argue this prop can be removed (triggers validation only on submit)